rpc: add logging tags for rpc connect events

knz opened the pull request:
Epic: CRDB-11517
First commit from #75237.

This ensures that the log messages sent to the DEV channel relating to
RPC low-level activity (e.g. conn failures, heartbeat etc) include the
remote node ID and address and connection class.

This also connects the heartbeat task to the tracing infra properly.

Example before:
```
[n6] 87  dialing n4: 127.0.0.1:41706 (default)
```

After:
```
[n6,rnode=4,raddr=127.0.0.1:41706,class=default] 87  dialing
```

knz commented on pkg/rpc/clock_offset.go:
```diff
@@ -172,7 +172,15 @@ func (r *RemoteClockMonitor) UpdateOffset(
 		}
-		latencyAvg.Add(float64(roundTripLatency.Nanoseconds()))
+
+		// If the roundtrip jumps by 50% beyond the previously recorded average, report it in logs.
```
I have attempted to look into CUSUM. I don't think that method is applicable, becasue we don't have an expected range of values beforehand.

knz commented:
@rauchenstein the next step for this PR is to implement a hysteresis (separate trhreshold for downward and upward gradients, computed as e.g. +/- 50% of last 30s mean) and see empirically if it delivers acceptable results.

knz reviewed (commented):
I like this. Thanks Jay.
One nit and I think this is good to go.

> Reviewed 2 of 3 files at r6, 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71243#-:-MrT3Snr7JiVdB-aJ44T:b-r8cr6r)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @shralex)

___
*[pkg/rpc/clock_offset.go, line 68 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/71243#-MrT3-bk1on0fWTIC-XH:-MrT3-blEaLEDnE3Dkxn:bgdevbb) ([raw file](https://github.com/cockroachdb/cockroach/blob/3547e8d7f5488a47aede5c6469e27b0635cdcb96/pkg/rpc/clock_offset.go#L68)):*
> ```Go
> func (t *resettingMaxTrigger) triggers(value, resetThreshold, triggerThreshold float64) bool {
> 	if *t {
> 		// "recently triggered" state
> ```

nit: here and below, make sentences and ensure capitals and final periods.
___
*[pkg/rpc/clock_offset.go, line 211 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/71243#-MrT3C1LEYEAu69wHy01:-MrT3C1LEYEAu69wHy02:b-6vj2vh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3547e8d7f5488a47aede5c6469e27b0635cdcb96/pkg/rpc/clock_offset.go#L211)):*
> ```Go
> 
> 		// If the roundtrip jumps by 50% beyond the previously recorded average, report it in logs.
> 		// Don't report it again until it falls belowe 40% above the average.
> ```

nit: below*
___
*[pkg/rpc/clock_offset_test.go, line 230 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/71243#-MrT3R9R25gdTKZP0ckm:-MrT3R9R25gdTKZP0ckn:bg60iqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3547e8d7f5488a47aede5c6469e27b0635cdcb96/pkg/rpc/clock_offset_test.go#L230)):*
> ```Go
> 
> func TestResettingMaxTrigger(t *testing.T) {
> 	defer leaktest.AfterTest(t)()
> ```

nice :100: 


<!-- Sent from Reviewable.io -->

rauchenstein reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71243)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @knz, and @shralex)

___
*[pkg/rpc/clock_offset.go, line 68 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/71243#-MrT3-bk1on0fWTIC-XH:-MrT8eTN-DvlsJvFAKjI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3547e8d7f5488a47aede5c6469e27b0635cdcb96/pkg/rpc/clock_offset.go#L68)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: here and below, make sentences and ensure capitals and final periods.
</blockquote></details>

Done.
___
*[pkg/rpc/clock_offset.go, line 211 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/71243#-MrT3C1LEYEAu69wHy01:-MrT8Y9o9RvolXdb9VLT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3547e8d7f5488a47aede5c6469e27b0635cdcb96/pkg/rpc/clock_offset.go#L211)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: below*
</blockquote></details>

Done.
___
*[pkg/rpc/clock_offset_test.go, line 230 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/71243#-MrT3R9R25gdTKZP0ckm:-MrTW-yO4mmJ1fhqUNzs:b9i3kuh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3547e8d7f5488a47aede5c6469e27b0635cdcb96/pkg/rpc/clock_offset_test.go#L230)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nice :100: 
</blockquote></details>

thanks. ack


<!-- Sent from Reviewable.io -->

knz commented:
you can rebase this on top of #74195 (which should be merged by the time you read this) to remove the first 1 commits from this PR and make it easier to review.

erikgrinaker commented on pkg/rpc/context.go:
```diff
@@ -1050,18 +1051,32 @@ type delayingHeader struct {
+	if remoteNodeID == 0 {
+		dialCtx = logtags.AddTag(dialCtx, "rnode?", nil)
+	} else {
+		dialCtx = logtags.AddTag(dialCtx, "rnode", remoteNodeID)
```
I think we should use the standard node formatting here, i.e. `n7` rather than `7`. We can maybe make this more compact by using e.g. `r=n7`.

erikgrinaker commented on pkg/rpc/context.go:
```diff
@@ -1050,18 +1051,32 @@ type delayingHeader struct {
+	} else {
+		dialCtx = logtags.AddTag(dialCtx, "rnode", remoteNodeID)
+	}
+	dialCtx = logtags.AddTag(dialCtx, "raddr", target)
```
This gets marked as sensitive and redacted in logs. Is that intentional? I'm not sure if we consider IP addresses to be sensitive, but it would make debugging a bit harder.

erikgrinaker reviewed (commented):
Thanks, this will be useful.

Please update the commit message and PR description. The title is a bit vague, and is identical to a previous PR/commit with the same title (which confused me at first since I mixed up this PR with the other one). Also, the message alludes to "two changes", but it only makes one change.

knz committed:
rpc: add a context arg to grpcDialRaw()

Release note: None

knz committed:
rpc: add context arg to grpcDialNodeInternal()

Release note: None

knz committed:
rpc: add a context arg to runHeartbeat()

Release note: None

knz committed:
rpc: annotate logs with remote addr, nodeID, class

This ensures that the log messages sent to the DEV channel relating to
RPC low-level activity (e.g. conn failures, heartbeat etc) include the
remote node ID and address and connection class.

Example before:
```
[n6] 87  dialing
```

After:
```
[n6,rnode=4,raddr=127.0.0.1:41706,class=default] 87  dialing
```

Release note: None

knz committed:
rpc: minor polish to log messages

This removes now-redundant logging output (since the same data is
included in tags) and adds a tag for the heartbeat background task.

For example, before:
```
[n6,rnode=4,raddr=127.0.0.1:41706,class=default] 87  dialing n4: 127.0.0.1:41706 (default)
```

After:
```
[n6,rnode=4,raddr=127.0.0.1:41706,class=default] 87  dialing
```

Release note: None

knz commented:
This is rebased and simplified.
PTAL

@rauchenstein FYI, the problem that we couldn't solve earlier, is that in `context_test.go`, the call to the `runHeartbeat()` method must be done using the `rpcCtx.masterCtx`, because that has a particular cancellation policy necessary for the HB task to terminate when the test terminates too.

andreimatei reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71243#-:-MttiEBU9GwkTDVFBd5o:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz and @shralex)

___
*[pkg/rpc/context.go, line 1149 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/71243#-Mttgfe0-a6ZJRRgduH9:-Mttgfe0-a6ZJRRgduHA:b4m6jqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/58e2f9e91dcead369f437f594393cadbdb288f0c/pkg/rpc/context.go#L1149)):*
> ```Go
> // that it respects the same cancellation policy.
> func (rpcCtx *Context) grpcDialRaw(
> 	dialCtx context.Context, target string, remoteNodeID roachpb.NodeID, class ConnectionClass,
> ```

you can't use `ctx` ?
___
*[pkg/rpc/context.go, line 1141 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/71243#-MtthUqI2jtRWNYiVDCv:-MtthUqJ6QnfY-uFHuTv:b-e778ej) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab2771407c1fb95bf761799c123e4a43348ca72b/pkg/rpc/context.go#L1141)):*
> ```Go
> 	dialCtx := rpcCtx.masterCtx
> 	if remoteNodeID == 0 {
> 		dialCtx = logtags.AddTag(dialCtx, "rnode?", nil)
> ```

maybe `rnode, '?'` ?
___
*[pkg/rpc/context.go, line 1302 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/71243#-Mtti4Qr96tI01ONlHPT:-Mtti4Qr96tI01ONlHPU:b7o8ii1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b48168c0d6ec04692ca888bcd0ea0199ea50de53/pkg/rpc/context.go#L1302)):*
> ```Go
> 		conn.grpcConn, redialChan, conn.dialErr = rpcCtx.grpcDialRaw(dialCtx, target, remoteNodeID, class)
> 		if conn.dialErr == nil {
> 			dialCtx = logtags.AddTag(dialCtx, "heartbeat", nil)
> ```

make this a new `ctx := ...` variable. I don't think you want to overwrite `dialCtx`.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71243)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @knz, and @shralex)

___
*[pkg/rpc/context.go, line 1149 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/71243#-Mttgfe0-a6ZJRRgduH9:-MtxhG-B1TfKp2oySJqs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/58e2f9e91dcead369f437f594393cadbdb288f0c/pkg/rpc/context.go#L1149)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

you can't use `ctx` ?
</blockquote></details>

Done.
___
*[pkg/rpc/context.go, line 1141 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/71243#-MtthUqI2jtRWNYiVDCv:-Mtxg54ICermc3GKBUe1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab2771407c1fb95bf761799c123e4a43348ca72b/pkg/rpc/context.go#L1141)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

maybe `rnode, '?'` ?
</blockquote></details>

Done.
___
*[pkg/rpc/context.go, line 1302 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/71243#-Mtti4Qr96tI01ONlHPT:-MtxhOTB4ga_NpzGnvFS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b48168c0d6ec04692ca888bcd0ea0199ea50de53/pkg/rpc/context.go#L1302)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

make this a new `ctx := ...` variable. I don't think you want to overwrite `dialCtx`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz commented:
TFYR!

bors r=andreimatei
